platform-hub-api/app/services/agents/keycloak_agent_service.rb

```diff
@@ -0,0 +1,141 @@
+            password: @password,
+            grant_type: 'password'
+          }
+      rescue => e
+        Rails.logger.error "Couldn't obtain access token from Keycloak: #{e.class} - #{e.message}"
+        raise
+      end
+
+      if response.success?
+        Rails.logger.info 'Obtained access token from Keycloak.'
+        @token = JSON.parse(response.body)['access_token']
+        @token_expires_at = JSON.parse(Base64.decode64(@token.split('.').second))['exp']
+        @token
+      else
+        Rails.logger.error "Couldn't obtain access token from Keycloak. Response: #{response.body}"
+        raise Errors::KeycloakAccessTokenRequestFailed
+      end
+    end
+
+    def get_user_representation(user)
+      keycloak_identity = user.identity(:keycloak)
+      user_id = keycloak_identity.try(:user_id)
+
+      raise Errors::KeycloakIdentityMissing if user_id.nil?
+
+      begin
+        response = client.get "/auth/admin/realms/#{@realm}/users/#{user_id}",
+          {},
+          {Authorization: "Bearer #{bearer_token}"}
+      rescue => e
+        Rails.logger.error "Couldn't get user representation from Keycloak: #{e.class} - #{e.message}"
+        raise
+      end
+
+      handle response
+
+      if response.success?
+        JSON.parse(response.body)
+      else
+        Rails.logger.error "Couldn't get user representation from Keycloak. Response: #{response.body}"
+        raise Errors::KeycloakUserRepresentationMissing
+      end
+    end
+
+    def update_user(representation = {})
+      user_id = representation['id']
+
+      if user_id.nil?
+        Rails.logger.error 'Keycloak user representation missing and will not be updated.'
+        raise Errors::KeycloakUserRepresentationMissing
+      end
+
+      begin
+        response = client.put "/auth/admin/realms/#{@realm}/users/#{user_id}",
+          representation.to_json,
+          {
+            'Authorization': "Bearer #{bearer_token}",
+            'Content-Type': 'application/json'
+          }
+      rescue => e
+        Rails.logger.error "Couldn't update user representation in Keycloak: #{e.class} - #{e.message}"
+        raise
+      end
+
+      handle response
+
+      if response.success?
+        Rails.logger.info 'User representation updated in Keycloak!'
+        representation
+      else
+        Rails.logger.error "User representation hasn't been updated in Keycloak! Response: #{response.body}"
+        nil
+      end
+    end
+
+    def deactivate_user(user)
+      update_enabled_state(user, false)
+    end
+
+    def activate_user(user)
+      update_enabled_state(user, true)
+    end
+
+    def update_enabled_state(user, state)
+      retries ||= 0
+      representation = get_user_representation(user)
+      representation['enabled'] = state
+      update_user(representation)
+    rescue Errors::KeycloakAccessTokenExpired
+      retry if (retries += 1) < 3
+    end
+
+    def token_not_expired?
+      remaining_time = token_expires_at - Time.now.to_i
```

Comment:
Is this a JWT token? If so, you can use the same JWT helpers as the auth concern.